src/core/tsi/ssl_transport_security_utils.cc

```diff
@@ -375,4 +380,79 @@ absl::StatusOr<std::string> AkidFromCrl(X509_CRL* crl) {
   return ret;
 }
 
+bool IsPemCertificateChainNonEmptyAndValid(absl::string_view cert_chain_pem) {
+  if (cert_chain_pem.empty()) {
+    return false;
+  }
+  BIO* in = BIO_new_mem_buf(cert_chain_pem.data(), cert_chain_pem.size());
+  if (in == nullptr) {
+    return false;
+  }
+  uint8_t* cert_data = nullptr;
+  long cert_length = 0;
+  size_t number_of_valid_certs = 0;
+  while (PEM_bytes_read_bio(&cert_data, &cert_length, /*pnm=*/nullptr,
+                            PEM_STRING_X509, in, /*cb=*/nullptr,
+                            /*u=*/nullptr)) {
+    // Try to form the X.509 object from the DER-encoding.
+    const uint8_t* der = cert_data;
+    X509* x509 = d2i_X509(/*out=*/nullptr, &der, cert_length);
```

Comment:
So, you're doing all the work to parse this, which then, presumably you in the server setup would hand this object to an SSL_CTX as use certificate to set up your context to start the server. 

It seems like this whole module is a bunch of extra work to do this twice instead of restructuring the code that starts the server up to do this, Fail if it does not work, and if it does, hand the X509 object to the SSL_ctx and start the server.  Why would you want to do this twice?

Reply:
Thanks @bob-beck! This is a great point. What I'm planning now is the following: in this PR, have the APIs for parsing the PEM; in a follow-up PR, call this API at server creation time and then plumb the parsed PEM down to where it needs to go. (This will be a lot of plumbing hence why I'd like to defer it to a follow-up.)